Explorer/Assets/Plugins/RustSegment/SegmentServerWrap/RustSegmentAnalyticsService.cs

```diff
@@ -0,0 +1,157 @@
+using DCL.PerformanceAndDiagnostics.Analytics;
+using Plugins.RustSegment.SegmentServerWrap.ContextSources;
+using Segment.Analytics;
+using Segment.Serialization;
+using System;
+using System.Collections.Generic;
+using UnityEngine.Pool;
+
+namespace Plugins.RustSegment.SegmentServerWrap
+{
+    /// <summary>
+    ///     This implementation is thread-safe
+    /// </summary>
+    public class RustSegmentAnalyticsService : IAnalyticsService
+    {
+        private enum Operation
+        {
+            Identify,
+            Track,
+            Flush,
+        }
+
+        private const string EMPTY_JSON = "{}";
+        private volatile string cachedUserId = string.Empty;
+        private readonly Dictionary<ulong, (Operation, List<MarshaledString>)> afterClean = new ();
+        private readonly IContextSource contextSource = new ContextSource();
+        private static volatile RustSegmentAnalyticsService? current;
+
+        public RustSegmentAnalyticsService(string writerKey)
+        {
+            if (string.IsNullOrWhiteSpace(writerKey))
+                throw new ArgumentNullException(nameof(writerKey), "Invalid key is null or empty");
+
+            if (current != null)
+                throw new Exception("Rust Segment previous instance is not disposed");
+
+            using var mWriterKey = new MarshaledString(writerKey);
+            bool result = NativeMethods.SegmentServerInitialize(mWriterKey.Ptr, Callback);
+
+            if (result == false)
+                throw new Exception("Rust Segment initialization failed");
+
+            ReportHub.Log(ReportData.UNSPECIFIED, "Rust Segment initialized");
+            current = this;
+        }
+
+        ~RustSegmentAnalyticsService()
+        {
+            current = null;
+            bool result = NativeMethods.SegmentServerDispose();
+
+            if (result == false)
+                throw new Exception("Rust Segment dispose failed");
+        }
+
+        public void Identify(string userId, JsonObject? traits = null)
+        {
+            lock (afterClean)
+            {
+                cachedUserId = userId;
+
+                var list = ListPool<MarshaledString>.Get()!;
+
+                var mUserId = new MarshaledString(userId);
+                var mTraits = new MarshaledString(traits?.ToString() ?? EMPTY_JSON);
+                var mContext = new MarshaledString(contextSource.ContextJson());
+
+                ulong operationId = NativeMethods.SegmentServerIdentify(mUserId.Ptr, mTraits.Ptr, mContext.Ptr);
+                AlertIfInvalid(operationId);
+
+                list.Add(mUserId);
+                list.Add(mTraits);
+                list.Add(mContext);
+
+                afterClean.Add(operationId, (Operation.Identify, list));
+            }
+        }
+
+        public void Track(string eventName, JsonObject? properties = null)
+        {
+            lock (afterClean)
+            {
+                var list = ListPool<MarshaledString>.Get()!;
```

Comment:
assertion for Identify being called could be as easy as just checking if `cachedId` is string null or empty. However, I would just logged error but still continue the flow (who know, maybe it will track it in some scenarios)

Reply:
On rust level it never tracks in a case if Identity wasn't called first